jni/src/Foreign/JNI.hs

```diff
@@ -475,28 +502,52 @@ getStaticMethodID cls methodname sig = withJNIEnv $ \env ->
     JNI.withString sig $ \sigp ->
     [CU.exp| jmethodID {
       (*$(JNIEnv *env))->GetStaticMethodID($(JNIEnv *env),
-                                           $(jclass cls),
+                                           $fptr-ptr:(jclass cls),
                                            $(char *methodnamep),
                                            $(char *sigp)) } |]
 
 getObjectClass :: Coercible o (J ty) => o -> IO JClass
 getObjectClass (coerce -> upcast -> obj) = withJNIEnv $ \env ->
+    objectFromPtr =<<
     [CU.exp| jclass {
       (*$(JNIEnv *env))->GetObjectClass($(JNIEnv *env),
-                                        $(jobject obj)) } |]
+                                        $fptr-ptr:(jobject obj)) } |]
 
+-- | Creates a global reference to the object referred to by
+-- the given reference.
+--
+-- Arranges for a finalizer to call 'deleteGlobalRef' when the
+-- global reference is no longer reachable on the Haskell side.
 newGlobalRef :: Coercible o (J ty) => o -> IO o
-newGlobalRef (coerce -> upcast -> obj) = withJNIEnv $ \env ->
-    fmap coerce $
-    [CU.exp| jobject {
-      (*$(JNIEnv *env))->NewGlobalRef($(JNIEnv *env),
-                                      $(jobject obj)) } |]
+newGlobalRef (coerce -> upcast -> obj) = withJNIEnv $ \env -> do
+    gobj <-
+      [CU.exp| jobject {
+        (*$(JNIEnv *env))->NewGlobalRef($(JNIEnv *env),
+                                        $fptr-ptr:(jobject obj)) } |]
+    coerce <$> J <$> newForeignPtrEnv finalize env gobj
+  where
+    finalize =
+      unsafePerformIO $ -- Memoize
+      withJNIEnv $ \env ->
+      [CU.exp| void (*)(JNIEnv *, jobject) { (*$(JNIEnv *env))->DeleteGlobalRef } |]
```

Comment:
This memoization is a hazard. Only the JNI environment of the thread calling `DeleteGlobalRef` can be used.
The GC might not always use the same thread to run the finalizers, and IIUC finalizers might even run concurrently.